<a href="https://colab.research.google.com/github/TPuHo4u/sal-rab-homeworks/blob/master/%D0%9A%D0%BE%D0%BF%D0%B8%D1%8F_%D0%B1%D0%BB%D0%BE%D0%BA%D0%BD%D0%BE%D1%82%D0%B0_%22SberMBA_Module1_Titanic_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Аналитика и предсказание выживаемости на Титанике


## План

1. Загрузка и базовое исследование данных

2. Более глубокое исследование данных

3. Задания для самостоятельной работы

4. Построение дашборда с аналитикой

5. Построение модели машинного обучения

Сегодня вам будет помогать AI-ассистент [GigaChat!](https://giga.chat) :)

## Подключение зависимостей / Загрузка библиотек

In [ ]:
import pandas as pd # для работы с таблицами
import numpy as np # для эффективных вычислений
import matplotlib.pyplot as plt
import seaborn as sns # для построения графиков

# 1. Загрузка и базовый анализ данных

Загрузим данные с помощью библиотеки ```pandas```, она же - ```pd``` в нашем примере. Для этого используем стандартный метод ```.read_csv(..)```, принимающий на вход путь к файлу с данными.


Загружаем данные о пассажирах Титаника по ссылке

In [ ]:
# Прямая ссылка на файл
url = "https://github.com/Murcha1990/datasets/raw/main/Titanic.xlsx"

# Загрузка данных
df = pd.read_excel(url, engine='openpyxl')

In [ ]:
# df = pd.read_excel("Titanic.xlsx")

Анализ данных и, вообще их исследование, начинается с того, что просто глазами смотрим на то, что находится внутри .csv файлов. Для этого достаточно вывести на экран первые несколько строк загруженных таблиц.

Метод ```.head()``` поможет нам в этом. С его помощью печатаем первые 5 строк из нашей таблицы:

In [ ]:
df.head()

Посмотрим на размер таблицы `df`:

In [ ]:
print(df.shape)

(891, 13)


В таблице 891 строка (то есть 891 пассажир) и 12 столбцов.

Количество строк в таблице еще можно посмотреть при помощи функии `len()`:

In [ ]:
len(df)

# 2. Более глубокий анализ данных

Оценим уровень выживаемости

In [ ]:
df['Survived']

In [ ]:
df['Survived'].mean() # доля выживших на Титанике

**Как влияет класс пассажира (Pclass) на выживаемость?**

Вычислим среднюю выживаемость для пассажиров Pclass=1, отдельно аналогично для Pclass=2 и для Pclass=3.

*Спросим у AI-ассистента!*

In [ ]:
df[...]

In [ ]:
df[df['Pclass'] == 1]

In [ ]:
# ответ AI-ассистента

In [ ]:
sns.barplot(x='Pclass', y='Survived', data=df);

In [ ]:
sns.barplot(x='Sex', y='Survived', data=df);

Более богатые выживали чаще.

**Как влияет пол на выживаемость?**

In [ ]:
df[df.Sex == "female"]['Survived'].mean(), df[df.Sex == "male"]['Survived'].mean()

Женщины выживали чаще!

**Как влияет порт посадки на выживаемость?**

Сначала посмотрим из каких портов садились на Титаник

In [ ]:
df['Embarked'].value_counts() # посмотрим, из каких портов садились пассажиры, и в каком количестве

In [ ]:
df[df.Embarked == 'S']['Survived'].mean(), df[df.Embarked == 'C']['Survived'].mean(), df[df.Embarked == 'Q']['Survived'].mean()

In [ ]:
sns.barplot(x='Embarked', y='Survived', data=df);

Можно строить и более интересные гистограммы!

In [ ]:
sns.barplot(x='Sex', y='Survived', hue='Pclass', data=df);

По результатам анализа мы выяснили, что выживаемость сильно зависит от пола, от класса пассажира, а также от порта посадки.

- Почти все женщины из 1го и 2го Класса спаслись.
- Жертвы серди женщин были в основном из 3 класса.
- Мужчины из 1 класса имеют больше шансов на спасение, чем из 2го и 3го классов.

Также наблюдается зависимость от порта посадки.

![](https://upload.wikimedia.org/wikipedia/commons/thumb/1/1f/%28ru%29_Titanic_voyage_map.svg/1920px-%28ru%29_Titanic_voyage_map.svg.png)

Вполне возможно, что это связано с загрузкой пассажиров, когда более поздние занимают места на более высоких палубах (аналогия - погрузка пассажиров в самолет), чтобы и центр масс был ниже и последующие пассажиры не мешалии не проходили мимо уже расположившихся.

# 3. Самостоятельная работа: анализ возраста пассажиров


Можно вычислять различные числовые характеристики столбцов.

Выше мы вычисляли среднее значение так:
`df['Survived'].mean() # доля выживших на Титанике`.

Аналогично, поставив после точки
* `max()`

* `min()`

* `median()`

можно вычислить максимум, минимум и медиану набора данных.

### Задание 1

Вычислите максимальный возраст пассажиров Титаника.

In [ ]:
# ваш код здесь

### Задание 2

Вычислите средний возраст пассажиров Титаника.

Затем вычислите средний возраст пассажиров только 1 класса.

Верно ли, что пассажиры 1 класса в среднем более возрастные, чем по всему Титанику?

In [ ]:
# ваш код здесь

### Задание 3

Создадим новый столбец `FamilySize`: размер семьи, путешествовавшей с пассажиром.

Сколько человек на Титанике путешествовали в семье размера 3?

In [ ]:
df['FamilySize'] = df['SibSp'] + df['Parch'] + 1

In [ ]:
# ваш код здесь

### Задание 4

Выведите на экран долю выживших для каждого размера семьи от 1 до 4 человек включительно. Верно ли, что чем больше размер семьи, тем выше вероятность выжить?

In [ ]:
 # ваш код здесь

## 4. Построение дашборда: Pandas profiling

In [ ]:
!pip install ydata_profiling -q

In [ ]:
from ydata_profiling import ProfileReport

In [ ]:
profile_report = ProfileReport(df, title='Pandas Profiling Report')

In [ ]:
profile_report

In [ ]:
profile_report.to_file('Titanic_report.html')

# 5. Машинное обучение

Сделаем следующие шаги:

* Сначала нужно разбить данные на тренировочную и тестовую части

* На тренировочных данных обучить модель

* На тестовых данных применить ее и получить прогнозы

* На тестовых данных оценить качество модели

Классические модели умеют работать только с числовыми признаками, поэтому:
1) закодируем колонки `Sex` и `Embarked` при помощи OneHotEncoding
2) Выбросим колонки `Name`, `Ticket`, `Cabin` и `PassengerId`, так как они не несут много полезной информации.

В результате в таблице df должны получиться только числовые колонки. Сохраните результат в таблицу `data`.

Попробуйте сделать это самостоятельно или при помощи AI-ассистента!

In [ ]:
# ваш код здесь

In [ ]:
data.head()

Сначала отделим признаки от целевой переменной:

In [ ]:
y = data['Survived'] # вектор целевой переменной

X = data.drop('Survived', axis = 1) # матрица объект-признак

In [ ]:
X.head()

Заполним пропуски в данных

In [ ]:
df.describe()

In [ ]:
mean_age = X['Age'].mean()

X['Age'] = X['Age'].fillna(mean_age)

При помощи функции `train_test_split` разобъем данные на тренировочную и тестовую части в пропорции 7 к 3.

In [ ]:
from sklearn.model_selection import train_test_split

Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.2, random_state=42) # задавая random_state, мы фиксируем разбиение на train и test - при каждом запуске ячейки оно теперь будет одинаковым

### Обучим модель и оценим ее качество

In [ ]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

model.fit(Xtrain, ytrain) # обучение модели на тренировочных данных

Применим модель к тестовым данным - для каждого пассажира из теста получаем 1, если модель считает, что он выжил, и 0 - если нет.

In [ ]:
pred = model.predict(Xtest)

pred

Посчитаем долю правильных ответов модели. Это можно сделать вручную, а можно использовать готовую функцию `accuracy` (доля правильных ответов):

In [ ]:
from sklearn.metrics import accuracy_score

accuracy_score(ytest, pred)

Модель верно определила, выживет или нет, почти 80% пассажиров Титаника!

## Бонус-соревнование: улучшение модели

Попробуйте придумать признаки, полезные для модели, используя столбец `Name`, а также извлечь полезные признаки и из других столбцов таблицы!

*Для получения идей и кода используйте AI-ассистента!*

In [ ]:
## Эти ячейки можно будет раскомментировать и запустить в случае, если вы поменяли таблицу непоправимо

# df = pd.read_excel("Titanic.xlsx") #, sheet_name=1)
# data = df.copy()

# data.drop(['Name', 'Ticket', 'Cabin', 'PassengerId'], axis=1, inplace=True)
# data = pd.get_dummies(data, columns=['Sex', 'Embarked'])

# data.head()

In [ ]:
# добавьте новые признаки в таблицу data
# Признаки можно извлечь из столбца Name в исходной таблице df и добавить в таблицу data

# ваш код здесь

Запустите ячейки ниже, чтобы потестировать вашу модель после добавления новых признаков!

In [ ]:
y = data['Survived'] # вектор целевой переменной

X = data.drop('Survived', axis = 1) # матрица объект-признак

mean_age = X['Age'].mean()

X['Age'] = X['Age'].fillna(mean_age)

In [ ]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, random_state=42) # задавая random_state, мы фиксируем разбиение на train и test - при каждом запуске ячейки оно теперь будет одинаковым

model = LogisticRegression()

model.fit(Xtrain, ytrain) # обучение модели на тренировочных данных

pred = model.predict(Xtest)

accuracy_score(ytest, pred)

## Полезные материалы

* [Курс по машинному обучению (если хотите углубиться в тему и продолжить свой путь в Machine Learning на серьезном уровне)](https://stepik.org/a/125501)